### PR TITLE
Feature/user roles and permission

### PR DESCRIPTION
**Pull request by IMMANUEL5015**
### **Description**
    This PR allows for the creation of user roles and permissions. Essentially, the fact that a user is logged in does not mean that they can just do whatever they like. Thus, this **Authorization** feature gives different levels of freedom to our app's users.

### **Overview**
      * An admin or regular user can create a bucketlist.
      * An admin user can get all bucketlists.
      * A regular user can get all of their own bucketlists.
      * An admin user can get any specific bucketlist.
      * A regular user can get any of their own specific bucketlist.
      * An admin user can update any specific bucketlist.
      * A regular user can update any of their own specific bucketlist.
      * An admin user can delete any specific bucketlist.
      * A regular user can delete any of their own specific  bucketlist.

### **Tradeoffs**
     Our app can register three types of users: 
       * admin
       * regular
       * invalid
   Normally, we wouldn't want an invalid user on our app. However, this was deemed necessary for the purpose of testing what happens when an invalid user (unauthorized user) tries to access a route that can be accessed only by an admin or a regular user.

### CODE REVIEW

**Comment by IMMANUEL5015** [comment]
If the .env file is put in .gitignore, TRAVISCI and COVERALLS will not be able to work with all the environment variables, which means that the tests will not work well.

I have tried it before. 

I will put the .env inside of .gitignore at the conclusion of the project. 

For now, I may have to work this way.

File: server/controllers/bucketlist.controller.js [hunk 1]
```diff
@@ -1,7 +1,22 @@
 const BucketList = require('../models/bucketlist.model');
 const User = require('../models/usermodel');
 
-//Create a new Bucketlist
+//Function for checking if a bucketlist is associated with a user
+function confirmDataAssociation(bucketlistId, userBucketlists){
+    let result = false;
+
+    //Loop through all the bucketlists belonging to the user
+    for(var i = 0; i < userBucketlists.length; i++){
```
Comment (ngozi-ekekwe) [inline server/controllers/bucketlist.controller.js]: Why do we need this method? I am thinking about performance here. if a user has 1000 bucket list items are we going to loop through 1000 items?

**Comment by ngozi-ekekwe** [comment]
> If the .env file is put in .gitignore, TRAVISCI and COVERALLS will not be able to work with all the environment variables, which means that the tests will not work well.
> 
> I have tried it before.
> 
> I will put the .env inside of .gitignore at the conclusion of the project.
> 
> For now, I may have to work this way.

I understand, but it's not safe to expose secret keys. Have you tried encrypting and decrypting during the build process? 

We can do this once we are done with the project

File: server/controllers/bucketlist.controller.js (hunk 1 above)
Comment (IMMANUEL5015) [reply server/controllers/bucketlist.controller.js]: I will enhance the performance of the application by ending the loop once we find the bucketlist. This means that we do not have to keep on looping even when our bucketlist has already been found.

Comment (IMMANUEL5015) [reply server/controllers/bucketlist.controller.js]: I have incorporated your valuable feedback and reduced the performance problem.

**Pull request merged**